README


This archive contains the distribution Padre-Plugin-XS,
version 0.09:

  Padre support for perl XS (and perlapi)

This software is copyright (c) 2010 by Steffen Mueller.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.




Changes
This is the change log for the Padre::Plugin::XS module:

0.10 2013-12-14
 - Added Dutch translation (ddn123456 - Dirk De Nijs)
 - Added Brazilian Portuguese translation (GARU)
 - Tweak pp-xs so that it will load under pp api 2.2 (BOWTIE)

0.09 2010-12-06
 - Restored missing POT/PO translation sources (Damyan Ivanov, AZAWAWI)

0.08 2010-10-29
 - Removed Test::NeedsDisplay dependency (AZAWAWI)

0.07 2010-10-27
 - [@PadrePlugin] replaced cruft in dist.ini (AZAWAWI)

0.06 2010-10-27
 - t/00-load.t is magically injected by [@PadrePlugin] via [LoadTests]
   (AZAWAWI)

0.05 2010-10-22
 - Re-uploaded 0.04 as 0.05 to satisfy the indexer (AZAWAWI)

0.04 2010-10-22
 - Switched to Dist::Zilla (AZAWAWI)
 - Fixed 00-load.t failure when $ENV{DISPLAY} is not set (AZAWAWI)

0.03 2010-10-20
 - Compatibility with the latest Padre (AZAWAWI, CHORNY)
 - Added i18n support (AZAWAWI)
 - Added Arabic translation (AZAWAWI)

0.02 2010-04-18
 - Bump dependency on Perl::APIReference for 5.12.0 support.

0.01 2009-11-21
 - First version unleashed on the unsuspecting world.


lib/Padre/Plugin/XS.pm
package Padre::Plugin::XS;

use 5.010001;
use warnings;
use strict;

use Padre::Unload;
use Try::Tiny;

our $VERSION = '0.10';
use parent qw(Padre::Plugin);

# Child modules we need to unload when disabled
use constant CHILDREN => qw{
	Padre::Plugin::XS
	Padre::Plugin::XS::PerlXS
	Padre::Plugin::XS::Document
	Perl::APIReference
};

#######
# Called by padre to know the plugin name
#######
sub plugin_name {
	return Wx::gettext('XS Support');
}

#######
# Called by padre to check the required interface
#######
sub padre_interfaces {
	return (
		'Padre::Plugin'   => '0.98',
		'Padre::Document' => '0.98',
		'Padre::Wx'       => '0.98',
		'Padre::Logger'   => '0.98',
	);
}


#########
# We need plugin_enable
# as we have an external dependency
#########
sub plugin_enable {
	my $self         = shift;
	my $perl_api_ref = 0;

	# Tests for externals used
	try {
		if ( require Perl::APIReference ) {
			$perl_api_ref = 1;
		}
	};

	return $perl_api_ref;
}

#######
# Add Plugin to Padre Menu
#######
sub menu_plugins_simple {
	my $self = shift;
	return $self->plugin_name => [
		Wx::gettext('About...') => sub { $self->plugin_about },
	];
}

#######
# Called by padre to know which document to register for this plugin
#######
sub registered_documents {
	return (
		'text/x-perlxs' => 'Padre::Plugin::XS::Document',
	);
}


#######
# Add icon to Plugin
#######
# Core plugins may reuse the page icon
sub plugin_icon {
	require Padre::Wx::Icon;
	Padre::Wx::Icon::find('logo');
	return;
}



#######
# plugin_about
#######
sub plugin_about {
	my $self = shift;

	# my $share = $self->plugin_directory_share or return;
	# my $file = File::Spec->catfile( $share, 'icons', '48x48', 'git.png' );
	# return unless -f $file;
	# return unless -r $file;

	my $info = Wx::AboutDialogInfo->new;

	# $info->SetIcon( Wx::Icon->new( $file, Wx::wxBITMAP_TYPE_PNG ) );
	$info->SetName('Padre::Plugin::XS');
	$info->SetVersion($VERSION);
	$info->SetDescription( Wx::gettext('Padre XS and perlapi support') );
	$info->SetCopyright('(c) 2008-2013 The Padre development team');
	$info->SetWebSite('http://padre.perlide.org/trac/wiki/PadrePluginXS');
	$info->AddDeveloper('Steffen Mueller <smueller@cpan.org>');
	$info->AddDeveloper('Ahmad M. Zawawi <ahmad.zawawi@gmail.com>');
	$info->AddDeveloper('Kevin Dawson <bowtie@cpan.org>');

	# $info->SetArtists(
	# [   'Scott Chacon <https://github.com/github/gitscm-next>',
	# 'Licence <http://creativecommons.org/licenses/by/3.0/>'
	# ]
	# );
	Wx::AboutBox($info);
	return;
}

########
# plugin_disable
########
sub plugin_disable {
	my $self = shift;

	# Close the dialog if it is hanging around
	$self->clean_dialog;

	# Unload all our child classes
	for my $package (CHILDREN) {
		require Padre::Unload;
		Padre::Unload->unload($package);
	}

	$self->SUPER::plugin_disable(@_);

	return 1;
}

########
# Composed Method clean_dialog
########
sub clean_dialog {
	my $self = shift;

	# Close the main dialog if it is hanging around
	if ( $self->{dialog} ) {
		$self->{dialog}->Hide;
		$self->{dialog}->Destroy;
		delete $self->{dialog};
	}

	return 1;
}

1;

__END__


=pod

=encoding UTF-8

=head1 NAME

Padre::Plugin::XS - Padre support for perl XS (and perlapi)

=head1 VERSION

version: 0.10

=head1 DESCRIPTION

This plugin is intended to extend Padre's support for editing XS
and C-using-perlapi.

=head1 SYNOPSIS
Currently the plugin implements limited syntax highlighting and
calltips using a configurable version of the perlapi. After installing
this plugin, you can enable XS calltips in the C<View> menu of Padre
and enjoy the full perlapi of various releases of perl while writing
XS code. You can configure the version of perlapi you write against in
the padre.yml of your project (key C<xs_calltips_perlapi_version>).
By default, the newest available perlapi will be used.

Once this plug-in is installed the user can switch the highlighting of
XS files to use the highlighter via the Preferences menu of L<Padre>.
The default XS syntax highlighting of Padre is abysmal. You're very
encouraged to enable the C<XS highlighter> feature.

To use this Plugin you need to make sure B<editor_calltips> is enabled in
Tools-Preferances-Advance

=head1 METHODS

=over 4

=item * clean_dialog

=item * menu_plugins_simple

=item * padre_interfaces

=item * plugin_about

=item * plugin_disable

=item * plugin_enable

=item * plugin_icon

=item * plugin_name

=item * registered_documents

=back


=head1 AUTHORS

Steffen Mueller E<lt>smueller@cpan.orgE<gt>

=head1 CONTRIBUTORS

Ahmad M. Zawawi E<lt>ahmad.zawawi@gmail.comE<gt>
Kevin Dawson E<lt>bowtie@cpan.orgE<gt>

=head1 ACKNOWLEDGMENTS

Many thanks to Gabor Szabo, who wrote the Kate plugin upon this is based.
I'm grateful to Herbert Breunung for writing Kephra and getting STC syntax highlighting more
right that us. Looking at his code has helped me write this.

=head1 COPYRIGHT

Copyright E<copy> 2010-2013 the Padre::Plugin::XS
L</AUTHOR> and L</CONTRIBUTORS> as listed above.


=head1 LICENSE

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.

=cut


lib/Padre/Plugin/XS/PerlXS.pm
package Padre::Plugin::XS::PerlXS;

use v5.10.1;
use warnings;
use strict;

use Padre::Current;
use Padre::Logger;
our $VERSION = '0.10';
use parent qw(Padre::Plugin);

sub colorize {
#	my $self = shift;
	TRACE("PerlXS colorize called") if DEBUG;
	# these are arguments: (maybe use for from/to?)
	#$current->editor->GetEndStyled,
	#$event->GetPosition

	my $doc = Padre::Current->document;

	#my $mime_type = $doc->get_mimetype;
	my $editor = $doc->editor;

	# TODO we might need not remove all the color, just from a certain section
	# TODO reuse the $first passed to the method
	#$doc->remove_color;

	my $c_keywords = join(
		' ', qw(
			auto break case char const continue default do
			double else enum extern float for goto if
			int long register return short signed sizeof static
			struct switch typedef union unsigned void volatile while
			)
	);
	my $pp_keywords = join(
		' ',
		'#define', '#include', '#if', '#ifdef', '#endif', '#undef',
		qw(
			__FILE__ __LINE__ __DATE__ __TIME__
			__STDC__ __STDC_VERSION__ __STDC_HOSTED__
			__cplusplus __OBJC__ __ASSEMBLER__
			__GNUC__ __COUNTER__ __GFORTRAN__
			__GNUC_MINOR__ __GNUC_PATCHLEVEL__
			__GNUG__ __STRICT_ANSI__ __ELF__ __VERSION__
			__OPTIMIZE__ __OPTIMIZE_SIZE__ __NO_INLINE__
			__GNUC_PATCHLEVEL__
			)
	);

	# TODO: there are lots more gnu specific pp keywords...

	my $perl_simple_types = join(
		' ', qw(
			I32 U32 STRLEN
			)
	);
	my $perlapi_structs = join(
		' ', qw(
			SV AV HV HE IV NV PV OP
			)
	);

	# TODO: Add colon where appropriate? Does that work at all?
	my $xs_keywords = join(
		' ', qw(
			MODULE PACKAGE ALIAS
			OUTPUT RETVAL
			CODE PPCODE PREFIX
			INIT NO_INIT PREINIT
			POSTCALL NO_OUTPUT CLEANUP
			INPUT SCOPE C_ARGS
			OUTLIST IN IN_OUTLIST IN_OUT
			BOOT REQUIRE VERSIONCHECK
			PROTOTYPES ENABLE DISABLE
			OVERLOAD FALLBACK INTERFACE
			INTERFACE_MACRO INCLUDE CASE
			THIS NO_INIT DESTROY
			)
	);

	my $c_keywords_docs = 'FIXME TODO';

	# TODO: cache
	my $perlapi_keywords = join( ' ', keys %{ $doc->keywords } );

#	$editor->SetLexer(Wx::Scintilla::Constant::SCLEX_CPP);
	$editor->SetProperty( "styling.within.preprocessor", 1 );

	# normal keywords (here: C)
	$editor->SetKeyWords( 0, $c_keywords . ' ' . $pp_keywords . ' ' . $perl_simple_types );

	# perlapi stuff and xs stuff
	$editor->SetKeyWords( 1, $perlapi_structs . ' ' . $xs_keywords . ' ' . $perlapi_keywords );
	$editor->SetKeyWords( 2, $c_keywords_docs );
	$editor->SetProperty( "braces.cpp.style", 10 );

	return ();
}




1;

__END__

=pod

=encoding UTF-8

=head1 NAME

Padre::Plugin::XS::PerlXS - previous code needs to be removed, but checked before hand.

=head1 VERSION

version: 0.10

=head1 METHODS

=over 4

=item * colorize


=back

=head1 AUTHOR

See L<Padre::Plugin::XS>

=head2 CONTRIBUTORS

See L<Padre::Plugin::XS>

=head1 COPYRIGHT

See L<Padre::Plugin::XS>

=head1 LICENSE

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.

=cut


lib/Padre/Plugin/XS/Document.pm
package Padre::Plugin::XS::Document;

use v5.10.1;
use strict;
use warnings;

use Padre::Logger;
use YAML::Tiny;

our $VERSION = '0.10';
use parent qw(Padre::Document);

#######
# Task Integration
#######
sub task_functions {
	return '';
}

sub task_outline {
	return '';
}

sub task_syntax {
#	return 'Padre::Plugin::YAML::Syntax';
	return '';
}

# sub comment_lines_str {
	# return '#';
# }



#######
# Padre-Document Overloading
#######

# TODO better highlighting. Can vim do better? Can we steal? Add an STC highlighter? ...

sub get_calltip_keywords {
	my $self = shift;
	TRACE("get_calltip_keywords called") if DEBUG;

	if ( not defined $self->{_perlapi_keywords} ) {
		$self->_load_perlapi_keywords();
	}
	return $self->{_perlapi_keywords};
}


# This loads the perlapi keywords for calltips
# It first tries to use the Perl::APIReference module which has perlapi references
# for many, many releases of perl. It fetches the desired perlapi version from the
# project configuration and uses the newest if not configured. Then, it asks the
# Perl::APIReference for the index of keywords. Since this can fail at various levels,
# we fall back to reading the perlapi keywords from an included YAML file if necessary
# --Steffen
sub _load_perlapi_keywords {
	my $self = shift;

	if ( not eval {use Perl::APIReference 0.03; 1;} ) {
		TRACE('Perl::APIReference failed') if DEBUG;
		$self->{_perlapi_keywords} =
			YAML::Tiny::LoadFile( Padre::Util::sharefile( 'languages', 'perl5', 'perl5.yml', 'perlapi_current.yml' ) );
		return;
	}

	my $perl_version = $self->{_perlapi_version};
# p $perl_version;
	if ( not defined $perl_version ) {
		my $project = $self->project;
		if ( defined $project ) {
			$perl_version = 'newest';
			my $cfg = $project->config;
			$perl_version = $cfg->xs_calltips_perlapi_version(); # this is = 'newest'
		} else {
			$perl_version = Padre->ide->config->xs_calltips_perlapi_version();
		}
		$self->{_perlapi_version} = $perl_version;
	}
# p $perl_version;
	my $apiref = eval { Perl::APIReference->new( perl_version => $perl_version ) };
# p $apiref;
	if ( not $apiref ) {

		# fallback...
		$self->{_perlapi_keywords} =
			YAML::Tiny::LoadFile( Padre::Util::sharefile( 'languages', 'perl5', 'perl5.yml', 'perlapi_current.yml' ) );
		return;
	}

	# TODO: Perl::APIReference also provides an "index" method, but that doesn't return the structure
	#       in exactly the way we want it. Easy way out: Add an accessor to Perl::APIReference that
	#       returns an API structure akin to what would be returned by loading the YAML.
#	require YAML::Tiny;
	$self->{_perlapi_keywords} = YAML::Tiny::Load( $apiref->as_yaml_calltips );
# p $self->{_perlapi_keywords};
	return;
}

1;

__END__

=pod

=encoding UTF-8

=head1 NAME

Padre::Plugin::XS::Document - A Padre Document that understands XS

=head1 VERSION

version: 0.10

=head1 METHODS

=over 4

=item * task_functions

=item * task_outline

=item * task_syntax

=item * get_calltip_keywords

=back

=head1 AUTHOR

See L<Padre::Plugin::XS>

=head2 CONTRIBUTORS

See L<Padre::Plugin::XS>

=head1 COPYRIGHT

See L<Padre::Plugin::XS>

=head1 LICENSE

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.

=cut


Build.PL

use strict;
use warnings;

use Module::Build 0.3601;


my %module_build_args = (
  'build_requires' => {
    'Locale::Msgfmt' => '0.15',
    'Module::Build' => '0.3601',
    'Test::More' => '0'
  },
  'configure_requires' => {
    'Module::Build' => '0.3601'
  },
  'dist_abstract' => 'Padre support for perl XS (and perlapi)',
  'dist_author' => [
    'Steffen Mueller <smueller@cpan.org>',
    'Ahmad M. Zawawi <ahmad.zawawi@gmail.com>'
  ],
  'dist_name' => 'Padre-Plugin-XS',
  'dist_version' => '0.09',
  'license' => 'perl',
  'module_name' => 'Padre::Plugin::XS',
  'recommends' => {},
  'recursive_test_files' => 1,
  'requires' => {
    'Padre' => '0.57',
    'Perl::APIReference' => '0.04'
  },
  'script_files' => [],
  'share_dir' => {
    'dist' => 'share'
  }
);


my $build = Module::Build->new(%module_build_args);

$build->create_build_script;
